docs/design/readme.md

```diff
@@ -0,0 +1,1704 @@
+
+- Provides DSDL (de)serialization by invoking Nunavut-generated code appropriately. This provides a higher-level API for the user code that operates on DSDL objects instead of raw binary blobs.
+- Demultiplexes incoming transfers across multiple consumers. This allows the application to instantiate multiple subscribers for a given subject simultaneously.
+- Provides a convenient interface for invoking and serving RPC service calls.
+
+**`libcyphal::application`** is a layer on top of the previous two that provides the top-level node-centric API for the application and implements some common functions of the protocol, such as the register interface, publication of heartbeats and port usage messages, and so on. This is where the included batteries can be found. Unlike the lower layers, the application layer depends on the code generated by Nunavut from certain standard DSDL definitions.
+
+## General considerations
+
+### Design principles
+
+Exceptions are not leveraged by the library itself, but it is designed to be exception-aware, allowing its integration into applications that make use of exception handling.
+
+Runtime type identification is not required but allowed as an optional feature to facilitate dynamic linking when used in a high-level operating system; more on this later.
+
+Usage of the heap is reduced to the bare minimum and is done strictly via polymorphic memory resources without any reliance on the global heap. Most allocations are done once during initialization; with the exceptions listed below, the library will not perform further allocations at runtime unless the application needs to change its port configuration (i.e., subscriptions/publications/RPC) dynamically at runtime, which is rarely seen in embedded real-time systems. The exceptions to the no-memory-allocation policy are as follows:
+
+- Transport-layer queues and memory reassembly buffers allocated at runtime by the lizards. The memory usage patterns and limits of the lizards are very well formalized in the documentation; further, certain lizards allow the replacement of heap allocators with block pool allocators.
+
+Type erasure is used throughout to reduce coupling and uphold encapsulation. In some cases, heap allocation that is often associated with type erasure is avoided with the help of small-object-optimized containers that provide an internal arena to store the type-erased object within. More on this below.
+
+Unconventionally, dynamic typing is leveraged with care in certain platform-specific contexts to uphold encapsulation.
+
+### RTTI and dynamic linking considerations
+
+As will be seen later, certain design features are dependent on the ability to detect identical types at runtime, yet the library must be usable without the RTTI capability of the C++ runtime. The following construct is used to provide a small subset of RTTI features that happens to be sufficient for the needs of the library:
+
+```c++
```

Comment:
>The benefit of this approach is that the GUIDs are all explicit identifiers visible in source and they do not change based on toolchain options

I can see that. A disadvantage is that it requires manual assignment of (pseudo) random type identifiers; one specific risk here (and a very real one, I suppose, considering how pervasive copy-paste errors are), which follows directly from the hands-on nature of this approach, is that a developer may forget to roll dice when defining a new type based on an existing one. In the other approach, the dice will be rolled by the linker.

Why can't we use uint64 for the type ID? If you have a hundred types, the collision probability assuming randomly chosen identifiers is negligible -- 2.68e-16.

Just for the sake of completeness, I should add that, assuming a reproducible build environment, one could derive the type ID automatically at compile time as a hash of the file name and the line number, although the disadvantages are clear.

Usage of a reserved value for `NoType` breaks the orthogonality of the type system; it should be an empty option instead.

Reply:
> requires manual assignment of (pseudo) random type identifiers

The science behind GUID generation is pretty tight and the probably of collision is considered effectively zero. Copy and paste errors are pernicious when using GUIDs. It is the largest issue when using this style of non-rtti programming. Overuse is also a large issue (see https://en.wikipedia.org/wiki/XPCOM) so the intent here, to allow type erasure of DSDL types and to enable the media layer interfaces to be abstract but introspectable, should be strictly adhered to. Perhaps this means we want to use both techniques? The COM-like approach for any externalize-able objects and the compile-time approach for internals?

> Why can't we use uint64 for the type ID?

Because of the "G" in GUID. (or "U" in UUID). See https://en.wikipedia.org/wiki/Universally_unique_identifier. GUIDs allow distribution of shared libraries in binary form whereas compile-time solutions require complete program recompilation. While this isn't a primary use-case the compile-time option specifically prohibits it. Furthermore, GUIDs can be logged and the log values can be found in a source search instead of requiring binary forensics.

> Usage of a reserved value for NoType breaks the orthogonality of the type system; it should be an empty option instead.

Perhaps I meant VoidType and would rename this.

Reply:
Moving to https://github.com/OpenCyphal/CETL/issues/51